xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/web/XWikiServletURLFactory.java

```diff
@@ -444,23 +447,97 @@ public URL createExternalURL(String spaces, String name, String action, String q
         return this.createURL(spaces, name, action, querystring, anchor, xwikidb, context);
     }
 
+    private void appendQueryParameter(String key, Object paramValue, StringBuilder stringBuilder)
+        throws UnsupportedEncodingException
+    {
+        if (paramValue instanceof String) {
+            stringBuilder.append(key);
+            stringBuilder.append("=");
+            stringBuilder.append(URLEncoder.encode((String) paramValue, "UTF-8"));
+        } else if (paramValue.getClass().isArray()) {
+            Class ofArray = paramValue.getClass().getComponentType();
+            if (ofArray.isPrimitive()) {
+                int length = Array.getLength(paramValue);
+                for (int i = 0; i < length; i++) {
+                    appendQueryParameter(key, Array.get(paramValue, i).toString(), stringBuilder);
+                    if (i < length - 1) {
+                        stringBuilder.append("&");
+                    }
+                }
+            } else {
+                Object[] zeArray = (Object[]) paramValue;
+                for (int i = 0; i < zeArray.length; i++) {
+                    appendQueryParameter(key, zeArray[i].toString(), stringBuilder);
+
+                    if (i < zeArray.length - 1) {
+                        stringBuilder.append("&");
+                    }
+                }
+            }
+        } else if (paramValue instanceof Collection) {
+            Collection zeCollection = (Collection) paramValue;
+            int index = 0;
+            for (Object paramValueElement : zeCollection) {
+                appendQueryParameter(key, paramValueElement.toString(), stringBuilder);
+
+                if (index < zeCollection.size() - 1) {
+                    stringBuilder.append("&");
+                }
+                index++;
+            }
+        } else {
+            appendQueryParameter(key, paramValue.toString(), stringBuilder);
+        }
+    }
+
+    private URL buildURL(URL serverUrl, String path, Map<String, Object> queryParameters, XWikiContext context)
```

Comment:
@tmortagne I had to entirely rewrite this one since we finally cannot rely on URIBuilder: it encodes directly the part of the URL, and that's something we also do in normalizeURL, so basically we encode it twice... I spot it with the failing test in XWikiServletFactoryURLTest that I updated.